sway/commands/keep_keyboard_layout.c

```diff
@@ -0,0 +1,24 @@
+#include "sway/config.h"
+#include "sway/commands.h"
+#include "sway/input/input-manager.h"
+#include "log.h"
+
+struct cmd_results *cmd_keep_keyboard_layout(int argc, char **argv) {
```

Comment:
I just realized that it would probably be better if this command was a seat command instead of a global one.

Reply:
@emersion thanks, can you describe what do I need to do to convert this command to be per-seat one? I'm not familiar with Sway codebase *(especially so half a year later since last time I worked with it)*.

Reply:
It should be pretty simple: remove it from the list of global commands, add it to the list of seat commands (see `commands/seat.c`), move this file to the `commands/seat` dir and rename it to have a `seat_` prefix. Then you can use it like this:

```
seat * keep_keyboard_layout per_window
```

You'll need to move the config field to the seat config too, so that different seats can have different settings.

Reply:
Ok, almost done, but how do I inside the

```c
static void seat_send_focus(struct sway_node *node, struct sway_seat *seat) {
```

the current `seat_config`. Do I have to add another argument, or can I somehow infer it from these two? `sway_seat` doesn't seem to have a pointer to its original `sway_config`.

Reply:
```c
struct seat_config *seat_config = seat_get_config(seat);  // May be NULL
```

Reply:
So, I converted the command to per-seat one *(didn't update docs yet, wanted to test first)*, but when I execute `swaymsg "seat * keep_keyboard_layout per_window"` it doesn't work for me. In case somebody wants to look at it and it's something trivial, [I pushed the changes to this branch](https://github.com/Hi-Angel/sway/commits/kbd-layout-per-window-seat).Myself I don't know when I gonna debug this.

Reply:
> I just realized that it would probably be better if this command was a seat command instead of a global one.

what would be the benefit of that?

Reply:
>what would be the benefit of that?

Better multi-seat support. Two seats of the same machine can have different settings.

Reply:
This feature is ready for a few months now. People want per window keyboard kbd layout.
I use 3 seats and exactly the same layouts - layout per window, not seat.
This PR is exactly about it - per window. Why not merge it?
Is there anyone using different kbd layouts per seat?

Reply:
@emersion thanks, so, I debugged it, and the problem turns out that [in this command handling function](https://github.com/Hi-Angel/sway/commit/20dd6077d9ad59de0a92a2eb4bbd9f09771833e6#diff-e2b76690ea4454a5495b846f716883deR6) the `seat_config` is not the one that is used when a user is switching between windows. Moreover: every time I execute `swaymsg "seat * keep_keyboard_layout per_window"`, the address of `config->handler_context.seat_config` in this function is different. It's like it's copied somewhere, though unlikely by me. Any ideas?

Reply:
Heh, I tried to use this as an opportunity to finally try reverse-execution, but when I stop `sway` in gdb, I'm getting spammed by `Key repeat discarded, Wayland compositor doesn't seem to be processing events fast enough!` until I continue. Oh, well.

Reply:
>thanks, so, I debugged it, and the problem turns out that in this command handling function the seat_config is not the one that is used when a user is switching between windows.

Yeah, configs are copied when applied iirc. But that shouldn't matter right? Since `keep_keyboard_layout` should be copied over?

Reply:
> Yeah, configs are copied when applied iirc. But that shouldn't matter right? Since `keep_keyboard_layout` should be copied over?

Thanks, right, but apparently I need to also modify the copying code to make this field copied too. Where is it copied?

Reply:
In `merge_seat_config`

Reply:
Thanks, I looked at it. Apparently to make this work I need to introduce an additional default enum member `KEYBOARD_LAYOUT_UNDEFINED`, because this function only copies over fields that were modified, and I need to figure out whether that happened.

I think instead of copying Sway should just apply functions that under `commands/seat/` dir directly to configs requested. In particular this would allow to get rid of "undefined" values.

Reply:
…and in general this would reduce maintainance because situations like mine just won't happen.

Reply:
`merge_seat_config` is needed for proper layering of wildcard and named seat configs. The seat configs are handled the way they are for a reason and the complexity wouldn't be there if it wasn't necessary.

Reply:
@RedSoxFan I don't see why the wildcard and named seat resolution couldn't be done just before applying the command, thus not doing any copying.

Reply:
Ignoring that the user can interlace wildcard and named seat commands, you would still need some way to merge those two configs together. Applying the wildcard config and then the named config will not result in desired behavior (especially if you are overriding the wildcard config with defaults).

Reply:
Correct me if I'm wrong, but what a user does is two things: 1. defines seat configs that it wants to apply a command to, and 2. provides name of the field and the exact value to set it to.

I don't really understand you, but even if you merge 2 configs where the value was set to what user wanted — you will get back a config with the value that a user have set.

Reply:
@emersion done. I also rebased on master. Note, there's a gotcha, I don't know if it's expected behavior: when I put in my config a line

```haskell
seat * keep_keyboard_layout per_window
```

it doesn't work for me. However this line does:

```haskell
exec swaymsg "seat * keep_keyboard_layout per_window"
```

Reply:
>I don't really understand you, but even if you merge 2 configs where the field was set to what user wanted — you will get back a config with the value that a user have set.

Merging `seat * option1 true` and `seat * option2 true` would result in `{"option1": false, "option2": true}` if merging the second one on top of the first one without "undefined" values.

There are more related issues, e.g. `seat * option1 true` + `seat seat1 option1 false`.

>when I put in my config a line it doesn't work for me

That's unexpected, other seat commands work in the config file.

Reply:
> > I don't really understand you, but even if you merge 2 configs where the field was set to what user wanted — you will get back a config with the value that a user have set.
>
> Merging `seat * option1 true` and `seat * option2 true` would result in `{"option1": false, "option2": true}` if merging the second one on top of the first one without "undefined" values.

But why would you merge them if these are separate commands? So you only get to merge `option1` with `option1` and `option2` with `option2`.

Either way, I think we just gone to a speculation, point of my text you quoted was that I don't understand why would they want to merge configs while executing `seat * option1 true` without doing the copy. You only merge the copy with existing configs, not multiple existing configs, right? So there should be nothing stopping one from executing command-handler functions directly on the existing configs.

> There are more related issues, e.g. `seat * option1 true` + `seat seat1 option1 false`.

These still end up resolved to something exact, i.e. still can be resolved before applying the command handlers, and the handlers can be applied to configs we resolved to.

> > when I put in my config a line it doesn't work for me
>
> That's unexpected, other seat commands work in the config file.

Any ideas?

Reply:
@emersion I figured the problem: I expected that the `seat_config` "copy" is created by `calloc()` as everything else in sway. Unfortunately instead it is created by `new_seat_config()` *(internally it still uses `calloc`, but that's not the point)*. So what happens is that I initialize `keep_keyboard_layout` in this function to a "valid" value; and then when upon `swaymsg` call or whatever the config copy is created *(to be acted upon and then to get merged with existing configs)*, it already has *valid* value.

Then the question is: where in Sway can I initialize default values of `seat_config`?

And while on it: I wonder, if it would be acceptable to leave a comment in `new_seat_config()` that it should not be used for assigning default values? So that future contributors won't get stuck on same problem with seat configs as I do.

Reply:
>where in Sway can I initialize default values of seat_config?

Sorry, I don't get it…

Reply:
> Sorry, I don't get it…

Let me reword it: upon creation of the config that Sway uses for the rest of its lifetime, I need to initialize a field to a default value. The config is created with `new_seat_config()`

The problem is: the "config copy" that this whole thread is about is also created with `new_seat_config()`, however the copy should have different default value, `UndefinedKbdLayout` *(or something similar, I'm talking from my memory)*.

My question is: where in code should I initialize a field of the first config to its default value so that this value won't get carried over to the "config copies" that sway creates.

Reply:
Why is this an issue that the "undefined keyboard layout" value gets carried over?

Reply:
> Why is it an issue that the "undefined keyboard layout" value gets carried over?

To be more explicit, "gets carried over" the value that I'd assign inside `new_seat_config()`.

If this is the "undefined keyboard layout", then if a user didn't explicitly set the option I'm implementing, on Sway startup the main config gonna have the "undefined keyboard layout" value which is a bug.

If this is a valid value, then, due to how Sway deals with config parsing *(i.e. creating copies and stuff)* I will have no way to differentiate "default value" from the one that user has explicitly set.

Reply:
Other options define a "not set" value (e.g. -1 if it's an int, or zero if it's an enum, etc). Set the field to this "not set" value in `new_seat_config`. Use this "not set" value to decide whether the existing value should be overridden when merging two configs. When actually reading the config value, handle "not set" as the value you want to be the default.

Is there an issue with that?

Reply:
>When actually reading the config value, handle "not set" as the value you want to be the default.
>
>Is there an issue with that?

Per my understanding, if one would set some irrelevant option with `swaymsg` command, then upon config merge mine option would have the -1. Now, if I treat this as "set option to a default value", this gonna override the value that a user could've previously set, which is a bug.

Reply:
No: in the merge function you override the current value only if it's != -1.

Reply:
> No: in the merge function you override the current value only if it's != -1.

This is not possible because I can't set it to -1 in `new_seat_config()`. To quote my prev. comment why:

> If this is the "undefined keyboard layout", then if a user didn't explicitly set the option I'm implementing, on Sway startup the main config gonna have the "undefined keyboard layout" value which is a bug.

Reply:
>the main config gonna have the "undefined keyboard layout" value which is a bug.

no it's not

Reply:
> > the main config gonna have the "undefined keyboard layout" value which is a bug.
> 
> no it's not

So, I just double-checked: in the latest code it requires one-line change: [replace here](https://github.com/swaywm/sway/pull/3155/commits/b1f665b1c3af6ba7f9923bbf41248909d537d934#diff-12676f2e76e92011b7450ecc7b4bdc1fR30) the `KEYBOARD_LAYOUT_GLOBAL` with `KEYBOARD_LAYOUT_UNDEFINED`.

What happens next is that if I remove from my configuration the line `seat * keep_keyboard_layout per_window`, sway crashes [on this sway_assert](https://github.com/swaywm/sway/pull/3155/commits/b1f665b1c3af6ba7f9923bbf41248909d537d934#diff-57d5973ad58d843753d353531e8cb85eR111), as I expected.

Reply:
>sway crashes on this sway_assert, as I expected.

Look, I really appreciate your work on this PR. However I'd like to optimize the time spent on the review. Both reviewer and contributor time is scarce, and I'm not a fan of endless discussions that go nowhere.

When I said "no it's not" I wasn't kidding: the code that reads values from the config does handle the "unset" case as if it was whatever-default-you-want (here, global). In this particular case your code checks for "per-window" so it should just work. But you're probably aware of this, so:

I can understand that you don't like how it's done, and I'm open to discuss them in separate issues. However it really shouldn't happen in this PR. The PR should just implement what it says in the title and blend with the existing code -- not rework the whole existing framework.

Let me know if something's unclear.

Reply:
> > sway crashes on this sway_assert, as I expected.
> 
> Look, I really appreciate your work on this PR. However I'd like to optimize the time spent on the review. Both reviewer and contributor time is scarce, and I'm not a fan of endless discussions that go nowhere.

I'm addressing your review. I'm assuming you know the code you're talking about, hence I ask you about how should I address your comments.

If you'd left the comments half a year ago when this PR was publicly worked on and finished, I'd probably did this research on my own. But right now I'm not using Sway, and to address your review I need collaboration from you.

> When I said "no it's not" I wasn't kidding: the code that reads values from the config does handle the "unset" case as if it was whatever-default-you-want (here, global). In this particular case your code checks for "per-window" so it should just work. But you're probably aware of this, so:
> 
> I can understand that you don't like how it's done, and I'm open to discuss that in separate issues. However it really shouldn't happen in this PR. The PR should just implement what it says in the title and blend with the existing code -- not rework the whole existing framework.
> 
> Let me know if something's unclear.

I don't know why are you thinking this, but I did **exactly** as you asked in prev. comment *(except I'm using a "undefined constant" instead of -1)*. Didn't I?

And yes, I also don't like that this discussion goes so long: this is the reason I linked you to exact code lines in my prev. comment, hoping that if you think that it should work whereas it clearly doesn't, you could maybe see in code why it doesn't.

I'm not trying to rework anything, at this point I don't care about sway code, I'm sincerely just trying to pass the review.

Reply:
>I don't know why are you thinking this, but I did exactly as you asked in prev. comment (except I'm using a "undefined constant" instead of -1). Didn't I?

Okay. Then remove the assert you've added, since it conflicts with:

>When I said "no it's not" I wasn't kidding: the code that reads values from the config does handle the "unset" case as if it was whatever-default-you-want (here, global). In this particular case your code checks for "per-window" so it should just work.

Reply:
> > I don't know why are you thinking this, but I did exactly as you asked in prev. comment (except I'm using a "undefined constant" instead of -1). Didn't I?
> 
> Okay. Then remove the assert you've added, since it conflicts with:
> 
> > When I said "no it's not" I wasn't kidding: the code that reads values from the config does handle the "unset" case as if it was whatever-default-you-want (here, global). In this particular case your code checks for "per-window" so it should just work.

As you wish. Done [in the latest update](https://github.com/swaywm/sway/compare/b1f665b1c3af6ba7f9923bbf41248909d537d934..cc02a58177c8f7271cbc1c1b741b43919454993d).

Reply:
Ah, wait, I forgot to add the update to assign undefined, let me do in a sec.

Reply:
Done

Reply:
@emersion I think everything is addressed. It works too, I tested. Any other comments?